### PR TITLE
rust: Fix link arguments

### PR DESCRIPTION
**Pull request by xclaesse**
Reduce code duplication by iterating target.get_dependencies() instead
of iterating target.link_targets and target.link_whole_targets
separately. This has the extra benefit of taking into account
transitive dependencies.
    
- For indirect C ABI static libraries, this adds missing "-l static="
  arguments.
- For indirect Rust ABI dependencies, this adds missing "-L" arguments.
- Verbatim modifier was used only for link_whole_targets, it's now
  always used when available.
    
Fixes: #11694

### CODE REVIEW

**Comment by xclaesse** [comment]
This is an alternative fix replacing https://github.com/mesonbuild/meson/pull/11719. I imported @sdroege's unit test.

File: mesonbuild/backend/ninjabackend.py [hunk 1]
```diff
@@ -1948,8 +1948,6 @@ def generate_rust_target(self, target: build.BuildTarget) -> None:
         args += ['--emit', f'dep-info={depfile}', '--emit', f'link={target_name}']
```
Comment (dnicolodi) [inline mesonbuild/backend/ninjabackend.py:1948]: `s/benefice/benefit/` in the commit message.

**Comment by xclaesse** [comment]
@dcbaker I think the CRT changes you made in https://github.com/mesonbuild/meson/pull/11902 never worked. You made a typo there: https://github.com/mesonbuild/meson/pull/11902/files#diff-76c8a26e81a7fd37cca2db144349e88dbcf9be56e32954ea722a9d106e4e400dR2049-R2050. This PR fix that typo, but then the CI fails with `error: could not find native static library `msvcrtd`, perhaps an -L flag is missing?`.

File: mesonbuild/backend/ninjabackend.py (hunk 1 above)
Comment (xclaesse) [reply mesonbuild/backend/ninjabackend.py:1948]: thx fixed

**Comment by xclaesse** [comment]
@dcbaker CRT issue has been worked around now, this PR is now ready for review.

**Review by sdroege — APPROVED** [review]

**Pull request merged**